File: Algorithms/023.MergeKSortedLists.cpp
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        if (lists.size() == 0) {
            return NULL;
        }
        int end = lists.size() - 1;
        while (end > 0) {
            int begin = 0;
            while (begin < end) {
                lists[begin] = mergeTwoLists(lists[begin], lists[end]);
                begin ++;
                end --;
            }
        }
        return lists[0];
    }
private:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode *dummyHead = new ListNode(1);
        ListNode *p = dummyHead;
        while (l1 != NULL && l2 != NULL) {
            if (l1->val < l2->val) {
                p->next = l1;
                l1 = l1->next;
            }
            else {
                p->next = l2;
                l2 = l2->next;
            }
            p = p->next;
        }
        if (l1 != NULL) {
            p->next = l1;
        }
        if (l2 != NULL) {
            p->next = l2;
        }
        return dummyHead->next;
    }
};

File: Algorithms/py/023.MergeKSortedLists.py
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        def _mergeTwoLists(l1, l2):
            dummyHead = ListNode(1)
            p = dummyHead
            while l1 and l2:
                if l1.val < l2.val:
                    p.next = l1
                    l1 = l1.next
                else:
                    p.next = l2
                    l2 = l2.next
                p = p.next
            if l1: 
                p.next = l1
            if l2:
                p.next = l2
            return dummyHead.next

        if not lists:
            return None
        end = len(lists) -1 
        while end:
            begin = 0
            while begin < end:
                lists[begin] = _mergeTwoLists(lists[begin], lists[end])
                begin += 1
                end -= 1
        return lists[0]

